chain/client/src/sync/state/mod.rs

```diff
@@ -0,0 +1,388 @@
+            chain_finalization_sender,
+            shard_syncs: HashMap::new(),
+        }
+    }
+
+    /// Apply a state sync message received from a peer.
+    pub fn apply_peer_message(
+        &self,
+        peer_id: PeerId,
+        shard_id: ShardId,
+        sync_hash: CryptoHash,
+        data: ShardStateSyncResponse,
+    ) -> Result<(), near_chain::Error> {
+        self.peer_source_state
+            .lock()
+            .unwrap()
+            .receive_peer_message(peer_id, shard_id, sync_hash, data)?;
+        Ok(())
+    }
+
+    /// Processes the requests that the state sync module needed the Chain for.
+    fn process_chain_requests(&mut self, chain: &mut Chain) {
+        while let Ok(request) = self.header_validation_queue.try_recv() {
+            let result =
+                chain.set_state_header(request.shard_id, request.sync_hash, request.header);
+            request.response_sender.send(result).ok();
+        }
+        while let Ok(request) = self.chain_finalization_queue.try_recv() {
+            let result = chain.set_state_finalize(request.shard_id, request.sync_hash);
+            request.response_sender.send(result).ok();
+        }
+    }
+
+    /// Main loop that should be called periodically.
+    pub fn run(
+        &mut self,
+        sync_hash: CryptoHash,
+        sync_status: &mut StateSyncStatus,
+        chain: &mut Chain,
+        highest_height_peers: &[HighestHeightPeerInfo],
+        tracking_shards: Vec<ShardId>,
+    ) -> Result<StateSyncResult, near_chain::Error> {
+        let _span =
+            tracing::debug_span!(target: "sync", "run_sync", sync_type = "StateSync").entered();
+        tracing::debug!(%sync_hash, ?tracking_shards, "syncing state");
+
+        self.peer_source_state.lock().unwrap().set_highest_peers(
+            highest_height_peers.iter().map(|info| info.peer_info.id.clone()).collect(),
+        );
+        self.process_chain_requests(chain);
+
+        let mut all_done = true;
+        for shard_id in &tracking_shards {
+            let key = (sync_hash, *shard_id);
+            let status = match self.shard_syncs.entry(key) {
+                Entry::Occupied(mut entry) => match entry.get_mut().result.try_recv() {
+                    Ok(result) => {
+                        entry.remove();
+                        if let Err(err) = result {
+                            tracing::error!(%shard_id, ?err, "State sync failed for shard");
+                            return Err(err);
+                        }
+                        ShardSyncStatus::StateSyncDone
+                    }
+                    Err(TryRecvError::Closed) => {
+                        return Err(near_chain::Error::Other(
+                            "Shard result channel somehow closed".to_owned(),
+                        ));
+                    }
+                    Err(TryRecvError::Empty) => entry.get().status(),
+                },
+                Entry::Vacant(entry) => {
```

Comment:
Could we encapsulate this into a function like `self.start_state_sync_for_shard` for better readability?

Reply:
There is already the run_state_sync_for_shard function and the stuff here really isn't any useful logic other than just boilerplate code.